src/main.cpp

```diff
@@ -302,10 +301,29 @@ void nimble_port_ll_task_func(void* args) {
 }
 }
 
+void calibrate_lf_clock_rc(nrf_drv_clock_evt_type_t event) {
+  // 16 * 0.25s = 4s calibration cycle
+  // Not recursive, call is deferred via internal calibration timer
+  nrf_drv_clock_calibration_start(16, calibrate_lf_clock_rc);
+}
+
 int main(void) {
   logger.Init();
 
   nrf_drv_clock_init();
+  nrf_drv_clock_lfclk_request(NULL);
+
+  // When loading the firmware via the Wasp-OS reloader-factory, which uses the used internal LF RC oscillator,
+  // the LF clock has to be explicitly restarted because InfiniTime uses the external crystal oscillator if available.
+  // If the clock is not restarted, the Bluetooth timers fail to initialize.
+  nrfx_clock_lfclk_start();
+  while (!nrf_clock_lf_is_running()) {
```

Comment:
This is an infinite loop. Is there any chance we never get out of this loop?

Reply:
The loop runs until the LF clock peripheral is confirmed to be running. Without this hardware peripheral, the long timers, watchdog, and Bluetooth stack won't work (on any device).

I don't know how this loop could fail to exit, except on bad or damaged silicon, I guess? In any case, this kind of busy waiting is even used in the NRF SDK, e.g. at https://github.com/cmdwtf/nRF5_SDK/blob/master/modules/nrfx/drivers/src/nrfx_clock.c#L252= .

If this loop is considered risky, I can implement a timeout, but it should not be necessary imho. In all my testing across three different devices this loop never failed.

Reply:
Thanks for the explanation! You're right : without clock, we cannot expect anything to work properly... And the watchdog will take care to reset the device!